src/S3IteratorLibsvm.cpp

```diff
@@ -0,0 +1,412 @@
+
+  return ret;
+}
+
+/**
+ * Moves index forward while data[index] is a space
+ * returns true if it ended on a digit, otherwise returns false
+ */
+bool S3IteratorLibsvm::ignoreSpaces(uint64_t& index, const std::string& data) {
+  while (isspace(data[index])) {
+    index++;
+  }
+  return isdigit(data[index]);
+}
+
+bool S3IteratorLibsvm::ignoreSpacesNotNewline(uint64_t& index,
+                                              const std::string& data) {
+  while (data[index] != '\n' && isspace(data[index])) {
+    index++;
+  }
+  return isdigit(data[index]);
+}
+
+template <class T>
+T S3IteratorLibsvm::readNum(uint64_t& index, std::string& data) {
+  if (!isdigit(data[index])) {
+    throw std::runtime_error("Error in the dataset");
+  }
+
+  uint64_t index_fw = index;
+  // if float or double also accept dots
+  if constexpr (std::is_same<T, double>::value ||
+                std::is_same<T, float>::value) {
+    while (isdigit(data[index_fw]) || data[index_fw] == '.') {
+      index_fw++;
+    }
+  } else {
+    while (isdigit(data[index_fw])) {
+      index_fw++;
+    }
+  }
+
+  char c = data[index_fw];
+  data[index_fw] = 0;
+
+  T result;
+  if constexpr (std::is_same<T, int>::value) {
+    sscanf(&data[index], "%d", &result);
+  } else if constexpr (std::is_same<T, double>::value) {
+    sscanf(&data[index], "%lf", &result);
+  } else if constexpr (std::is_same<T, float>::value) {
+    sscanf(&data[index], "%f", &result);
+  } else if constexpr (std::is_same<T, uint64_t>::value) {
+    sscanf(&data[index], "%lu", &result);
+  } else {
+    throw std::runtime_error("Data type not supported");
+  }
+
+  data[index_fw] = c;  // repair
+  index = index_fw;
+  return result;
+}
+
+bool S3IteratorLibsvm::buildDatasetCsv(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+bool S3IteratorLibsvm::buildDatasetVowpalWabbit(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+/**
+ * Build minibatch from text in libsvm format
+ * We assume index is at a start of a line
+ * index moves forward
+ */
+bool S3IteratorLibsvm::buildDatasetLibsvm(
+    std::string& data,
+    uint64_t& index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+#ifdef DEBUG
+  std::cout << "building dataset libsvm index: " << index << std::endl;
+#endif
+  // libsvm format
+  // <label> <index1>:<value1> <index2>:<value2>
+
+  try {
+    std::vector<std::vector<std::pair<int, FEATURE_TYPE>>> samples;
+    std::vector<FEATURE_TYPE> labels;
+
+    samples.resize(minibatch_rows);
+    labels.resize(minibatch_rows);
+
+    for (uint64_t sample = 0; sample < minibatch_rows; ++sample) {
+      // ignore spaces
+      if (!ignoreSpaces(index, data)) {
+#ifdef DEBUG
+        std::cout << "ignoreSpaces did not find dig index: " << index
+                  << std::endl;
+#endif
+        // did not end up in a digit
+        return false;
```

Comment:
Shouldn't this throw an std::runtime_error stating that there is something wrong with the dataset? IgnoreSpaces should always end on a digit.

Reply:
There is also the case when the last sample is cut off, that case is not because the dataset is wrong.

But I'm going to add some code here to check whether the ignoreSpaces returns `false` because we found a sample that is partially cut or because the sample has a wrong format.